ESPS/general/man/man3/findespsfi.3
.\" Copyright (c) 1993 Entropic Research Laboratory, Inc.; All rights reserved
.\" @(#)findespsfi.3	1.10 24 Sep 1997 ERL
.TH FIND_ESPS_FILE 3\-Eu 24 Sep 1997
.SH NAME
find_esps_file \- finds full path to an ESPS or waves+ file.
.SH SYNOPSIS
.nf
#include <esps/esps.h>

char *find_esps_file(fullpath,filename,defpath,env_var_name)
char* fullpath;
char* filename; 
char *defpath;
char *env_var_name
.fi
.SH DESCRIPTION
.PP
Looks along a path for \fIfilename\fR. If the file is found and is
readable then return the the string containing the full Unix path to
the file.  If the file is not found NULL is returned.
.PP
Environment variables may also be used in the requested file.
Variable \fBESPS_BASE\fR is treated specially (see below).  An example
of this would be:
.nf

   find_esps_file(fullpath,"bin/$ARCH/xwaves",defpath,env_var_name);

.fi

.PP
The parameter \fIfilename\fR should the basename of the file to be
discovered. If \fIfilename\fR already specifies a full path (begins
with / or ./), then that path is returned (if it points to a readable
file). The parameter \fIfullpath\fR
should be a string large enough to hold the longest path on your
system, or NULL in which case a new string will be allocated and
returned containing the file path.  You should deallocate the string
when you do not need it anymore. Parameter \fIfilename\fR (provided or a
newly allocated) is returned.
.PP
The first two paramaters may be the same string.  However, remember
that it is \fBVERY\fR important that the string be large enough to
hold the resulting path. 
.PP
Parameter \fIdefpath\fR specifies the default search path.  The
syntax of the search path (just like Unix) is given below.
.PP
The default search path can be overridden by the environment variable
specified in \fBenv_var_name\fR.  If the environment variable specified
by \fBenv_var_name\fR is not set, or cannot be parsed then
the default file path (as specified by \fIdefpath\fR ) is used.
.PP
The syntax of the search path is a colon separated list of
directories; just like the normal Unix path. An example path is:
.PP
.nf

   $ESPS_BASE/bin:/usr/local/bin/$ARCH:/usr/esps/demo:$HOME/esps/bin

.fi
Environment variables may be used in the path by prefixing them
with a $ sign.  The value of environment variable \fBESPS_BASE\fR is
retrieved by calling \fIget_esps_base\fR(3\-\s-1ESPSu\s+1) . If the
environment variable is not set then it is replaced in the path
with nothing.
.SH EXAMPLES
.nf

/* provide string  */

char datafile[255];
(void) find_esps_file(datafile,"speech.sd","/usr/esps/bin:/usr/esps/demo","ESPS_PATH"); 

/* allocate new string */

char *datafile;
datafile  = find_esps_file(NULL,"speech.sd","/usr/esps/bin:/usr/esps/demo","ESPS_PATH"); 

.fi
.SH CONVIENENCE MACROS
.PP
Several macros are provided in include file $ESPS_BASE/include/esps/epaths.h.  They provide standard environment variables to specify paths and
standard default paths if the standard environment variable is not
set.
.PP
Parameter \fIa\fR is the string to fill the path in (or NULL)
(corresponds find_esps_file(3) parameter \fIfullpath\fR).
Paramater \fIb\fR is the file desired (corresponds to \fIfind_esps_file\fR(3)
parameter \fBfilename\fI).
.PP
FIND_ESPS_MENU(a,b) \- find ESPS menu files.  
.PP
FIND_ESPS_BIN(a,b)  \- find ESPS executable files.
.PP
IFIND_ESPS_PARAM_FILE(a,b)  \- find ESPS paramater files.
.PP
FIND_ESPS_FILTER(a,b) \- find ESPS filter files.
.PP
FIND_ESPS_LIB(a,b)  \- search for files in the ESPS library.
.PP
FIND_ESPS_INPUT(a,b) \- search for ESPS input files.
.PP
FIND_WAVES_INPUT(a,b) \- search for waves+ input files.
.PP
FIND_WAVES_LIB(a,b) \- search for files in the waves+ library
directory.
.PP
FIND_WAVES_COMMAND(a,b) \- find waves+ command files.
.PP
FIND_WAVES_MENU(a,b) \- find waves+ menu files.
.PP
FIND_WAVES_COLORMAP(a,b) \-  find waves+ colormaps.
.PP
FIND_WAVES_PROFILE(a,b) \- find a waves profile .
.PP
FIND_FAB2_BIN(a,b) \- find FAB2 board binaries.
.PP
FIND_SURF_BIN(a,b) \- find Surf board binaries.
.PP
.SH ERRORS AND DIAGNOSTICS
.PP
None.  Always returns something unless too small a string is provided
which could cause problems if other memory addresses are overwritten.
.SH FUTURE CHANGES
.PP
Make environment variable handling match the way the shell handles
them. Things like ${foobar}, etc... .
.SH BUGS
.PP
If the user provided string is too small then who knows what will happen.
.SH REFERENCES
.PP
None.
.SH "SEE ALSO"
.PP
.nf
\fIget_esps_base\fP(3\-\s-1ESPSu\s+1), \fIget_esps_base\fP(1\-\s-1ESPS\s+1), 
\fIfind_esps_file\fP(1\-\s-1ESPS\s+1), \fIbuild_filename\fP(3\-\s-1ESPS\s+1)
.fi
.SH AUTHOR
Ken Nelson wrote the routine and the man page.
.PP



















ESPS/general/src/lib/findespsfi.c
/*
 * This material contains unpublished, proprietary software of 
 * Entropic Research Laboratory, Inc. Any reproduction, distribution, 
 * or publication of this work must be authorized in writing by Entropic 
 * Research Laboratory, Inc., and must bear the notice: 
 *
 *    "Copyright (c) 1990-1991  Entropic Research Laboratory, Inc. 
 *                   All rights reserved"
 *
 * The copyright notice above does not evidence any actual or intended 
 * publication of this source code.     
 *
 * Written by:  Kenneth C. Nelson
 * Checked by:
 * Revised by:
 *
 * Brief description: Search for file along a path given as arg or env variable
 *
 */

static char *sccs_id = "@(#)findespsfi.c	1.11	4/6/93	ERL";

/* INCLUDE FILES */

#include <stdio.h>
#include <string.h>
#if !defined(LINUX_OR_MAC)
#include <esps/unix.h>
#endif
#include <esps/esps.h>
#include <sys/types.h>
#if !defined(APOLLO_68K) && !defined(OS5)
#include <sys/dir.h>
#else
#include <dirent.h>
#endif
#ifndef MAXNAMLEN
# define MAXNAMLEN 255
#endif

/* EXTERNAL VARIABLE DECLARATIONS */

extern int debug_level;  /* Provided by user level programs */
extern char *savestring(); /* ESPS savestring routine */

/* LOCAL FUNCTION DECLARATIONS */

/* 
 * fileReadable() - Local function to check and see if the file is readable
 *
 */

static int fileReadable(name)
char *name;
{
 FILE *file;
 int result;

  file = fopen(name,"r");
  if (file != NULL)
  {
    result = TRUE;
    fclose(file);
  }
  else
  {
    result = FALSE;
  }
  
 return result;

}


/*
 * expandEnvVar() - expands and environment variable
 *                  treats ESPS_BASE specially.
 *
 */

static void expandEnvVar(into_string,envvar)
 char *into_string;
 char *envvar;
{

   /* Use get_esps_base(3) to find value for ESPS_BASE */

   if (strcmp(envvar,"ESPS_BASE") == 0)
   {
     get_esps_base(into_string);
   }
   else  /* get a normal ESPS value */
   {
     if (getenv(envvar) != NULL)
     {
       strcpy(into_string,getenv(envvar));
     }
     else
     {
       strcpy(into_string,"");
     }
   }
}

/*
 * build_filename() - local routine to build up a file name
 *                   given a string to build in, a filename,
 *                   and a path component which can contain
 *                   environment variables to expand.
 *
 */

char *build_filename(into_string,filename,dirname)
char *into_string;
char *filename;
char *dirname;

{

 char envvar[MAXNAMLEN];    /* holds name of env var if found */
 char envexp[MAXNAMLEN];    /* holds env var expansion */
 char pathpart[MAXNAMLEN];  /* holds part part of dirname if found */
 char charstr[2];     /* Place to put, chars in */
 int  envmode = FALSE;
 char *ch;

 ch = dirname;
 strcpy(charstr," ");  /* Build up a 1 character string */
 strcpy(into_string,"");

 while (*ch != '\0')
 {

   switch (*ch)
   {
     case '$':
       if (envmode)
       {
	 expandEnvVar(envexp,envvar);
         strcat(into_string,envexp);
       }
       envmode = TRUE;
       strcpy(envvar,"");
       break;

      case '/':        /* things that terminate paths */
      if (envmode)
       {
	 expandEnvVar(envexp,envvar);
         strcat(into_string,envexp);
       }
       envmode = FALSE;
       strcpy(envvar,"");
       if (*ch == '/')  /* Add the slash */
       {
         charstr[0] = *ch;
         strcat(into_string,charstr);  
       }
       break;

      default:
        charstr[0] = *ch;
        if (envmode)
        {
          strcat(envvar,charstr);
        }
	else
        {
	  strcat(into_string,charstr);
        }
        break;
     }
  ch++ ;
} 
 /* Catch any pesky env vars at the end of the path */

 if (envmode) /* If an enviroment var was being built when the string ended*/
 {
    expandEnvVar(envexp,envvar);
    strcat(into_string,envexp);
 } 

/* Add the file being searched for if one was provided. */

 if (strlen(filename) > 0)
 {
   strcat(into_string,"/");
   strcat(into_string,filename);
 }

 return into_string;
}


/* LOCAL CONSTANTS */

#define EC_SCCS_DATE "4/6/93"
#define EC_SCCS_VERSION "1.11"

/* LOCAL MACROS */

/* Separator in path strings */

#define PATH_FS ":"


/*
 * find_esps_file() - finds a file along a path. Returns full path to
 *                    file if it is readable.
 */

char *find_esps_file(fullpath,filename,defpath,env_var_name)
char* fullpath;
char* filename;
char *defpath;
char *env_var_name;

{

 char *searchPath;   /* Holds the list of paths to search */
 char *currentDir;  /* Holds filepath being checked currently */
 char currentFile[MAXNAMLEN];
 char thePath[MAXNAMLEN];
 char expandedFilename[MAXNAMLEN];  /* holds requested file after ENV var
				     *  expansion.
				     */
 char *returnedString = NULL; 

 int  found = FALSE;

 /* Expand ENV vars in the requested file */

   build_filename(expandedFilename,"",filename);

 /* Check to see if it is a full path, if so then return it */
 
if ( (expandedFilename[0] == '/') || 
      (expandedFilename[0] == '.' && expandedFilename[1] == '/'))
{
    if (fileReadable(expandedFilename))
    {
      found = TRUE;
      if (fullpath == NULL)
       returnedString = savestring(expandedFilename);
      else
      {
        strcpy(fullpath,expandedFilename);
	returnedString =  fullpath;
      }
    }
}
else  /* full path not specified already */
{
 /* First build path string (from default or from env variable. */

 /* If env_var_name specifies a path then use it. */

  if (env_var_name != NULL && getenv(env_var_name) != NULL)
  {
     searchPath = savestring(getenv(env_var_name)); /* use environment var */
  }
  else /* use the default path */
  {
     spsassert(defpath != NULL,"find_esps_file: default path is NULL");
     searchPath = savestring(defpath);
  }


 /* Now for each path field separator, loop and try to find the file */
 
  found = FALSE;
  currentDir = strtok(searchPath,PATH_FS);
  while (currentDir != NULL && !found)
  {    
     build_filename(currentFile,expandedFilename,currentDir);
     found = (fileReadable(currentFile) != NULL);
     currentDir = strtok(NULL,PATH_FS); /* Go to next field */
  }

 free(searchPath);

/* Now if we found it return it else return NULL */

 if (found)
 {
   if (fullpath == NULL) 
   {
     returnedString = savestring(currentFile);
   }
   else
   {
     strcpy(fullpath,currentFile);
     returnedString =  fullpath;
   }
 }
 else
 {
    returnedString = NULL;
 }
}


if (found)
{
  if (debug_level > 1)
  {
    fprintf(stderr,"find_esps_file: %s found as %s\n",filename,returnedString);
  }  
}
else
{
  if (debug_level > 1)
  {
     fprintf(stderr,"find_esps_file: %s not found\n",filename);
  }
}

return returnedString;

}
